### PR TITLE
6: async checkpoints

### PR DESCRIPTION
**Pull request by jakkdl**
I had an eye on this one when working on trio103/104, which is why that one is a bit overkill for a relatively simple check - but made it a breeze to implement this one since it's got a large overlap.

Added messages temporarily as 300 & 301 to avoid numbering conflicts, will change that to 1XX once ready to merge. (Unfortunately needs changing in a bunch of places in the code)

### CODE REVIEW

**Review by Zac-HD — COMMENTED** [review]
Looks good!

File: tests/trio300_301.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+_ = ""
+
+
+async def foo():
+    await foo()
+
+
+async def foo2():  # error
+    ...
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply tests/trio300_301.py:14]: handy that there's already a `has_decorator` helper function ^^

**Comment by jakkdl** [comment]
rebased onto master, fixed your comments, fixed TRIO103 errors in the 107_108 file, and incremented version~ :sparkles:

**Review by Zac-HD — APPROVED** [review]
✨ 🎉 🚢

**Pull request merged**